transform.ml

```diff
@@ -1,205 +1,103 @@
-           * if its a goto (then we already remove it -- see above) or if its a branch (which
-           * is excluded by the second tests "succ (pc-1) = [pc]")
-           * TODO: we should implement some generic api for instructions like, "Instr.is_jmp" *)
-        remove_empty_jmp pc' acc
-      | i, _ ->
-        remove_empty_jmp pc' (i::acc)
-  in
-  remove_empty_jmp 0 []
+(* types of transformations *)
+type transform_instructions = instruction_stream -> instruction_stream option
+type create_optimistic_version = afunction -> version option
+type opt_function = afunction -> afunction option
+type opt_prog = program -> program option
 
-let remove_unreachable_code (instrs : instruction_stream) : instruction_stream =
-  let unreachable =
-    let merge _ _ _ = None in
-    let update _ _ = () in
-    Analysis.forward_analysis () instrs merge update
-  in
-  let rec remove_unreachable pc acc =
-    let pc' = pc+1 in
-    if pc = Array.length instrs then
-      Array.of_list (List.rev acc)
-    else
-      let i = instrs.(pc) in
-      match unreachable pc with
-      | exception Analysis.UnreachableCode _ ->
-        remove_unreachable pc' acc
-      | () ->
-        remove_unreachable pc' (i::acc)
-  in
-  remove_unreachable 0 []
+(* combining transformations *)
+let (&&&) t2 t1 instrs =
+  match t1 instrs with
+  | None -> t2 instrs
+  | Some instrs ->
+    begin match t2 instrs with
+      | None -> Some instrs
+      | Some instrs -> Some instrs
+    end
 
-let branch_prune (func : afunction) : afunction =
-  let version = Instr.active_version func in
-  let instrs = version.instrs in
-  let scope = Scope.infer func version in
-  let live = Analysis.live instrs in
-  let rec branch_prune pc acc =
-    if pc = Array.length instrs then
-      Array.of_list (List.rev acc)
-    else
-      match scope.(pc) with
-      | DeadScope -> assert(false)
-      | Scope scope ->
-        let pc' = pc + 1 in
-        begin match[@warning "-4"] instrs.(pc) with
-        | Branch (exp, l1, l2) ->
-          let osr = List.map (function
-              | (Const_var, x) ->
-                Osr_const (x, (Simple (Var x)))
-              | (Mut_var, x) ->
-                if List.mem x (live pc) then
-                  Osr_mut (x, x)
-                else
-                  Osr_mut_undef x)
-              (ModedVarSet.elements scope)
-          in
-          branch_prune pc'
-            (Goto l2 :: Osr (exp, func.name, version.label, l1, osr) :: acc)
-        | i ->
-          branch_prune pc' (i::acc)
-        end
-  in
-  let final = branch_prune 0 [] in
-  let cleanup = {
-    label = Rename.fresh_version_label func version.label;
-    instrs = remove_empty_jmp (remove_unreachable_code final);
-    annotations = None } in
-  { func with
-    body = cleanup :: func.body }
+let (!!) t instrs =
+  match t instrs with
+  | None -> false, instrs
+  | Some instrs -> true, instrs
 
+let (!!!) t instrs =
+  let _, instrs = !! t instrs in instrs
```

Comment:
I'd rather avoid prefix operators used in a non-local way (shortening stuff within a single function is fine, but otherwise...). What about `try_apply : ('a -> 'a option) -> 'a -> 'a`?